hoomd/pytest/test_local_snapshot.py

```diff
@@ -0,0 +1,579 @@
+"""Test that `hoomd.data.LocalSnapshot` and hoomd.data.LocalSnapshotGPU` work.
+"""
+from copy import deepcopy
+import hoomd
+from hoomd.array import HOOMDGPUArray
+import numpy as np
+import pytest
+try:
+    # This try block is purely to allow testing locally without mpi4py. We could
+    # require it for testing, and simplify the logic here. The CI containers all
+    # have mpi4py.
+    from mpi4py import MPI
+except ImportError:
+    skip_mpi4py = True
+else:
+    skip_mpi4py = False
+
+skip_mpi4py = pytest.mark.skipif(
+    skip_mpi4py, reason='mpi4py could not be imported.')
+
+try:
+    # We use the CUPY_IMPORTED variable to allow for local GPU testing without
+    # CuPy installed. This code could be simplified to only work with CuPy, by
+    # requiring its installation for testing. The CI containers already have
+    # CuPy installed when build for the GPU.
+    import cupy
+    CUPY_IMPORTED = True
+except ImportError:
+    CUPY_IMPORTED = False
+
+
+"""
+_N and _types are distinct in that the local snapshot does not know about them.
+We use the underscore to signify this. Those keys are skipped when testing the
+local snapshots, though are still used to define the state.
+"""
+
+Np = 5
+_particle_data = dict(
```

Comment:
There isn't really a reason to underscore all the variable names, is there? I see the comment above for the dict members, but the dictionaries themselves don't need it.

Reply:
It was underscored as a way to signify that it shouldn't be used directly for tests, that the fixtures created are what should be used. I have seen that pattern before. We can change it, however.

Reply:
In general, if you've seen things before feel free to push back, no need to just change things when I suggest them if there's nothing wrong. I'm fine with leaving it if you think it's a useful pattern.